OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs

```diff
@@ -458,7 +471,23 @@ public void Tick(PlayerResources pr)
 
 			var costThisFrame = RemainingCost / RemainingTime;
 			if (costThisFrame != 0 && !pr.TakeCash(costThisFrame))
+			{
+				if (!insufficientFundsPlayed)
+				{
+					insufficientFundsPlayed = true;
+					foreach (var funds in insufficientFunds)
+						funds.InsufficientFunds(owner.PlayerActor);
+				}
+
 				return;
+			}
+
+			if (insufficientFundsPlayed)
+			{
+				insufficientFundsPlayed = false;
+				foreach (var funds in insufficientFunds)
+					funds.SufficientFunds(owner.PlayerActor);
```

Comment:
You need to move the `foreach` out of this block. Otherwise you won't get a warning in all circumstances. Ie. if you have $300 left and try to build a helipad, you'll get a warning. If you then build a E1 the build succeeds, which should reset the warning. However, if you try to build an E3 after the E1, the warning will not appear. Moving the `foreach` will fix that.